src/raftstore/coprocessor/split_observer.rs

```
@@ -12,59 +12,49 @@
+    fn on_split(&mut self, _: &mut ObserverContext, split: &mut SplitRequest) -> Result<()> {
         if !split.has_split_key() {
             return Err("split key is expected!".to_owned());
         }
-        let split_key = split.get_split_key();
-        bytes::decode_bytes(split_key)
-            .map_err(|e| format!("invalid key {:?}: {}", split_key, e))
-            .map(|r| r.0)
-    }
-
-    fn on_split(&mut self, _: &mut ObserverContext, split: &mut SplitRequest) -> Result<()> {
-        let key = try!(self.extract_sql_key(split));
-
-        // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle or
-        // TABLE_PREFIX + table_id + TABLE_INDEX_MARK
+        let mut key = split.get_split_key().to_vec();
 
-        if !key.starts_with(TABLE_PREFIX) {
-            // if it's not a data key, just split.
+        if key.len() < VERSION_LENGTH {
+            // bypass non-mvcc key.
             return Ok(());
         }
-        // table id is a u64 and we only care about row data.
-        let mark_index = TABLE_PREFIX.len() + 8;
-        let new_key: &[u8];
-        if key.len() > mark_index && key[mark_index..].starts_with(TABLE_ROW_MARK) {
-            // handle is a u64 too.
-            let expected_len = TABLE_PREFIX.len() + 8 + TABLE_ROW_MARK.len() + 8;
-            if key.len() < expected_len {
-                return Err(format!("invalid key {:?}; len should be at least {}",
-                                   key,
-                                   expected_len));
-            }
-            new_key = &key[..expected_len];
+
+        // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle + column_id
+        // + version or TABLE_PREFIX + table_id + TABLE_INDEX_MARK + index_id + values + version
+        // or meta_key + version
+
+        let table_prefix_len = table::TABLE_PREFIX.len() + table::ID_LEN;
+        if key.starts_with(table::TABLE_PREFIX) && key.len() > table::PREFIX_LEN + table::ID_LEN &&
```

It's better to throw an error and cancel split when the key starts with TABLE_PREFIX but has a length shorter than it should be.